### PR TITLE
Nit: Combine App::state and App::userState to fix test races

### PR DESCRIPTION
**Pull request by oskirby**
## Description
Spotted in the wild: a flaky [testSubscription.js](https://github.com/mozilla-mobile/mozilla-vpn-client/actions/runs/9281903149/job/25539154427?pr=9617).

This was hauntingly familliar to a bug I fixed in one of my experimental hackathons in which `vpn.authenticateInApp()` returns success as soon as we get to the `UserAuthenticated` state but the model initialization is still waiting on a bunch of tasks to finish. If this happens during the initial test setup, and we quit before those jobs finish -  we are left with a settings file that is incomplete (the token is set but no keys are) and all the subsequent test jobs fail to load.

To fix this properly, I think the best approach is to combine the `App::state` and `App::userState` states together to eliminate potential race conditions between them, and to add a new `userAuthenticated` property that returns `true` once the auth is completed.

## Reference
Cherry-picking a bunch of stuff from PR #9584
Flaky test run: [testSubscription.js](https://github.com/mozilla-mobile/mozilla-vpn-client/actions/runs/9281903149/job/25539154427?pr=9617)

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

**Comment by oskirby** [comment]
This is confusing having both a `State` and `UserState` that both kind of do the same thing - perhaps a bigger refactoring is in order here. Is this a legacy of the attempted split-app suite thing?

File: src/app.cpp [hunk 1]
```diff
@@ -36,23 +36,14 @@ App::~App() { MZ_COUNT_DTOR(App); }
 
 int App::state() const { return m_state; }
 
-App::UserState App::userState() const { return m_userState; }
-
 void App::setState(int state) {
   logger.debug() << "Set state:" << state;
 
   m_state = state;
+  emit userAuthenticationChanged();
```
Comment (mcleinman) [inline src/app.cpp:43]: Why is this emitted here? Isn't it possible the user state didn't change?

**Review by mcleinman — CHANGES_REQUESTED** [review]
I really, really like this change. Thank you!

One question on a change. Thanks!

File: src/app.cpp (hunk 1 above)
Comment (oskirby) [reply src/app.cpp:43]: There is no longer a `userState` to change, but the `userAuthenticated` property is derived from the state now, so when the state changes we need to emit a notification that `userAuthenticated` may have changed value as well.

Comment (mcleinman) [reply src/app.cpp:43]: Agreed fully, though I missed that we check if user is authenticated in the one place we listen to that signal.

I think we should rename this to `userAuthenticationMaybeChanged` or something like that - to make clear that's its not guaranteed that the state changed when that is emitted.

Comment (oskirby) [reply src/app.cpp]: Done!

**Review by mcleinman — APPROVED** [review]
Thanks, looks great!

**Pull request merged**